### ISSUE TITLE
OpenSSL 1.1.x: static inline issue with Solaris Studio compiler (and GCC)

### ISSUE THREAD

**Issue by makr** [openssl/openssl#6912]
Hi OpenSSL-Team,

while porting an internal library to the OpenSSL 1.1 API, we are facing an issue when trying to build a debug-version (-g) of our library. The library itself loads libcrypto via dlopen(), only when it is actually needed. But in the sources we do of course include the OpenSSL-headers. When building our library without optimization and with debug-symbols (-g) the linker complains about undefined symbols. It turns out these are all from ...
* the openssl/lhash.h macro definition DEFINE_LHASH_OF, if openssl/ssl.h is included, or
* the openssl/safestack.h macro definition SKM_DEFINE_STACK_OF, if just openssl/crypto.h is included.

I constructed a small example C source (inline.c) and will attach it to this issue. To see the errors you may compile this with GCC, like ...
```
$ gcc -std=c99 -fkeep-inline-functions -g -o inline -I/path/to/openssl-1.1/include -ldl inline.c
/tmp/ccL4XtZc.o: In function `sk_OPENSSL_STRING_num':
/path/to/openssl-1.1/include/openssl/safestack.h:151: undefined reference to `OPENSSL_sk_num'
[...]
/tmp/ccL4XtZc.o: In function `sk_void_set_cmp_func':
/path/to/openssl-1.1/include/openssl/crypto.h:94: undefined reference to `OPENSSL_sk_set_cmp_func'
collect2: error: ld returned 1 exit status
```
... or the Oracle Developer Studio 12.5 compiler (Linux) ...
```
$ c99 -g -o inline -I/path/to/openssl-1.1/include -ldl inline.c
inline.o: In function `sk_OPENSSL_STRING_num':
inline.c:(.text+0x31): undefined reference to `OPENSSL_sk_num'
[...]
inline.o: In function `sk_void_set_cmp_func':
inline.c:(.text+0x1c29): undefined reference to `OPENSSL_sk_set_cmp_func'
```
... or Oracle Developer Studio 12.5 compiler (Sparc Solaris) ...
```
$ c99 -g -o inline-sparc -I/path/to/openssl-1.1/include inline.c
Undefined                       first referenced
 symbol                             in file
OPENSSL_sk_pop_free                 inline.o
[...]
OPENSSL_sk_find_ex                  inline.o
ld: fatal: symbol referencing errors
```

We have a solution on Linux, where we use GCC. There it is enough to just remove ```-fkeep-inline-functions```. Although this switch was probably added for a reason.

On Solaris, however, we have to use the Studio compiler. There is no way (we know of) to instruct this compiler to drop unused static functions other than to switch on the optimizer. This is, however, not acceptable for the debug-version of our library.

Do you have any suggestions or ideas, how to get around this?

Thanks!
-- Matthias Kraft

**Comment by makr** [comment]
I am somehow unable to attach this file, so here are the few lines:
```C
#include <stdio.h>
#include <dlfcn.h>

#include <openssl/crypto.h>

const char* (* fOpenSSL_version) (int) = NULL;

#define CALL_F(F) (f##F)
#define LOAD_F(F) if ((load_fct( ossl_h, #F, (void*)&f##F )) == 0) return 2

int load_fct( void *h, const char *fn, void **fp ) {
    if ((*fp = dlsym( h, fn )) == NULL)
        return 0;
    return 1;
}

int main(int argc, char *argv[]) {
    void *ossl_h;
    if ((ossl_h = dlopen( "libcrypto.so.1.1", (RTLD_GLOBAL|RTLD_LAZY) )) == NULL)
        return 1;
    LOAD_F(OpenSSL_version);
    printf( "%s\n", CALL_F(OpenSSL_version)( OPENSSL_VERSION ) );
    dlclose( ossl_h );
    return 0;
}
```

**Comment by dot-asm** [comment]
Example doesn't have to be that elaborated,
```
#include <openssl/crypto.h>
int main(){}
```
is sufficient...

**Comment by dot-asm** [comment]
What you apparently can do in mean time is to compose .h file of own and list those unresolved function as following:
```
#pragma weak OPENSSL_sk_num
...
```

**Comment by makr** [comment]
Thanks! This gets us going again.

**Comment by dot-asm** [comment]
On note related to your user identity, Matthias, and not to this problem. You had AIX-specific modifications to dso_dlfcn.c that allow to load modules from .a. Could you resubmit it as separate request (for master)?

**Comment by makr** [comment]
Yes, I still have that on my todo list. I am short on time, however. Will have look to expedite things and push it through it test cycle next week ...

**Comment by levitte** [comment]
Is this issue still current or should it be closed?

**Comment by levitte** [comment]
(I've labeled it 'after 1.1.1', because I don't think this is important enough to fix for the 1.1.1 release)

**Comment by makr** [comment]
I understand that you want to get the version out. Just to make it obvious for everyone looking at this issue: The mentioned workaround means you have to create a header file with a line
```#pragma weak <symbol>```
for every symbol mentioned by the compiler and include that header file after
```#include <openssl/crypto.h>```

**Comment by makr** [comment]
Now, having some time to work with OpenSSL again, I come back to this issue. Actually I come back here, because I believe that this issue now makes it impossible to build a debug version of OpenSSL 1.1.1a on a Solaris box :
```
19:10:48 cc  -Iinclude -xarch=generic64 -xstrconst -Xa -g -D_REENTRANT   -c -o test/rsa_complex.o test/rsa_complex.c
19:10:48 cc: Warning: -xarch=generic64 is deprecated, use -m64 to create 64-bit programs
19:10:49 /usr/ccs/bin/makedepend -f- -Y --  -Iinclude -xarch=generic64 -xstrconst -Xa -g -D_REENTRANT  -- test/rsa_complex.c 2>/dev/null \
19:10:49     > test/rsa_complex.d
19:10:49 rm -f test/rsa_complex
19:10:49 ${LDCMD:-cc} -xarch=generic64 -xstrconst -Xa -g -L. -mt -R '$ORIGIN/../lib'  -R '/opt/.../openssl/lib' \
19:10:49 	-o test/rsa_complex test/rsa_complex.o \
19:10:49 	 -lsocket -lnsl -ldl -lpthread -lm
19:10:49 cc: Warning: -xarch=generic64 is deprecated, use -m64 to create 64-bit programs
19:10:49 Undefined			first referenced
19:10:49  symbol  			    in file
19:10:49 OPENSSL_sk_pop_free                 test/rsa_complex.o
19:10:49 OPENSSL_sk_dup                      test/rsa_complex.o
19:10:49 OPENSSL_sk_pop                      test/rsa_complex.o
19:10:49 OPENSSL_sk_num                      test/rsa_complex.o
19:10:49 OPENSSL_sk_new                      test/rsa_complex.o
19:10:49 OPENSSL_sk_set                      test/rsa_complex.o
19:10:49 OPENSSL_sk_free                     test/rsa_complex.o
19:10:49 OPENSSL_sk_find                     test/rsa_complex.o
19:10:49 OPENSSL_sk_push                     test/rsa_complex.o
19:10:49 OPENSSL_sk_sort                     test/rsa_complex.o
19:10:49 OPENSSL_sk_zero                     test/rsa_complex.o
19:10:49 OPENSSL_sk_is_sorted                test/rsa_complex.o
19:10:49 OPENSSL_sk_shift                    test/rsa_complex.o
19:10:49 OPENSSL_sk_value                    test/rsa_complex.o
19:10:49 OPENSSL_sk_delete_ptr               test/rsa_complex.o
19:10:49 OPENSSL_sk_unshift                  test/rsa_complex.o
19:10:49 OPENSSL_sk_new_null                 test/rsa_complex.o
19:10:49 OPENSSL_sk_set_cmp_func             test/rsa_complex.o
19:10:49 OPENSSL_sk_reserve                  test/rsa_complex.o
19:10:49 OPENSSL_sk_new_reserve              test/rsa_complex.o
19:10:49 OPENSSL_sk_delete                   test/rsa_complex.o
19:10:49 OPENSSL_sk_insert                   test/rsa_complex.o
19:10:49 OPENSSL_sk_deep_copy                test/rsa_complex.o
19:10:49 OPENSSL_sk_find_ex                  test/rsa_complex.o
19:10:49 ld: fatal: symbol referencing errors
19:10:49 *** Error code 2
19:10:49 make: Fatal error: Command failed for target `test/rsa_complex'
19:10:49 Current working directory /.../build/openssl-1.1.1a
19:10:49 *** Error code 1
19:10:49 make: Fatal error: Command failed for target `all'
```

**Comment by mspncp** [comment]
Just a side note: The issue title has become a little bit outdated by now. Would you mind removing the `OpenSSL 1.1.0h:` prefix, or replacing it by `OpenSSL 1.1.x:` or `OpenSSL 1.1:` (whichever you prefer)?

**Comment by blastwave** [comment]
This is a very real problem with Oracle Studio 12.6 and Solaris 10 systems. I am seeing the exact same problem and it is a show stopper. There seems to be no way around this mess.

**Comment by blastwave** [comment]
I don't see what the purpose of this source file is : 

`/*
 * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
 *
 * Licensed under the OpenSSL license (the "License").  You may not use
 * this file except in compliance with the License.  You can obtain a copy
 * in the file LICENSE in the source distribution or at
 * https://www.openssl.org/source/license.html
 */

/*
 * Check to see if there is a conflict between complex.h and openssl/rsa.h.
 * The former defines "I" as a macro and earlier versions of the latter use
 * for function arguments.
 */
#if defined(__STDC_VERSION__)
# if __STDC_VERSION__ >= 199901L
#  include <complex.h>
# endif
#endif
#include <openssl/rsa.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
    /* There are explicitly no run time checks for this one */
    return EXIT_SUCCESS;
}
`

**Comment by mspncp** [comment]
Note: @blastwave, I added some triple backquotes ( \`\`\` ) around your code to have it displayed and indented properly. You can even add a 'C'  resp. 'diff' after the starting triple backquote ( \`\`\`C resp. \`\`\`diff ) to enable syntax highlighting for the C code resp. the diff output.

**Comment by mspncp** [comment]
> Special note .. the github markdown or whatever is needed here for trivial and readable comments is trash.  Certainly the OpenSSL project can have a real bugzilla similar to Apache ?

De gustibus non est disputandum ...

**Comment by makr** [comment]
Incidentally I am also coming back to this issue today. I need to make TLS 1.3 available for our products now. As the unpatched source of OpenSSL 1.1.1a is not compilable on Solaris I'd like to have some guidance, how to proceed. Is PR #7721, perhaps with the change suggested by @kroeckx, the way to go? Would it be integrated into the sources, so I do not need to maintain a patch?

**Comment by blastwave** [comment]
Hello Matthias, have no fear I think there is just a few trivial tweaks needed to get OpenSSL 1.1.1a building and testing fine on Solaris 10.  I am using Oracle Studio 12.6 and also C99 strict and due to various oddities such as NI_MAXHOST and NI_MAXSERV being formally undefined in RFC3493 we can generally get a nice result with a few little changes. I have had TLS 1.3 working well for quite a while now in Apache trunk 2.5.1 and am working on httpd 2.4.38 now.  That explains my sudden interest here.  I can email you a fairly specific set of instructions that should get the job done.

**Comment by mspncp** [comment]
> Hello Matthias, have no fear  I think there is just a few trivial tweaks needed to get OpenSSL 1.1.1a building and testing fine on Solaris 10.

@blastwave I am not the right person to address (although I have no fear ;-) ), because I neither have the hardware nor the  framework available to do some building or testing. The last time I used a Sun Workstation was way back during my time at  the university. But if you have any bugfixes or improvements for OpenSSL on Solaris 10, please don't hesitate to open a pull request, it will be welcome.

**Comment by makr** [comment]
Some more background can be found in the discussion of PR #7721 .

**Comment by mendobar** [comment]
I have built the openssl binary executable using openssl-1.1.1c on a Solaris 64-bit system with an AMD processor, using the Sun Studio C 12.4 C compiler.

The configuration was:

./Configure solaris64-x86_64-cc -static no-ec no-mdc2 no-rc5 no-idea no-asm -fPIC --openssldir=<path-sanitized-for-web>

The 1.1.1c files lhash.h and safestack.h already contain the pragma shown here.

I still get the following error.

$ ./openssl
ld.so.1: openssl: fatal: relocation error: file openssl: symbol OPENSSL_sk_new_null: referenced symbol not found
Killed
$ 

Is there a code change I can make that will eliminate this?
I have already looked at about a half a dozen other pages in this forum and tried rearranging link order and playing with compiler options to no avail.





$ ./openssl 
ld.so.1: openssl: fatal: relocation error: file openssl: symbol OPENSSL_sk_new_null: referenced symbol not found
Killed
$
